scalableminds/webknossos
# Add UI to restore previous versions of a tracing

This PR adds UI to restore older versions of a skeleton tracing. Restoring an older version simply creates a "new version" in the version history of the tracing - so it is possible to revert the reversion :)
The version preview and restore reuses functionality that we implemented for the task hot swap.

![version-restore-highlight](https://user-images.githubusercontent.com/1702075/45428378-6842d380-b6a1-11e8-88c2-e4ffcd762cd5.png)


### ToDo:
- [x] @philippotto In the version restore view, when scrolling, I would like to keep the headline on top and only scroll the version entries. However, I'm not css/html-savy enough to accomplish this. Any ideas?
- [x] @philippotto I'm having trouble with the flow types for the updateActions that are sent by the server. The server sends the same updateActions that we are sending, BUT there's an additional `actionTimestamp` (and maybe `version`). I tried to use flows $Call syntax to adapt our types, but the resulting flow type looks wrong to me (when looking at it through the type-at-pos functionality). I also can't get rid of the remaining flow errors.
- [x] @fm3 The versions are acquired through the datastore `updateActionLog` route, which returns batches of update actions. However, those batches are missing a version number. Currently the frontend is computing those versions, assuming the last entry is version 1, and counting up. Ideally, the backend would send the version number for each batch.
- [ ] We need to decide whether we want to group subsequent `createNode` actions by time interval or so, to avoid having too many of those `createNode` actions spamming the version restore view.

### URL of deployed dev instance (used for testing):
- https://restoreoldversions.webknossos.xyz

### Steps to test:
- Test the version restore mode
  - Create a skeleton tracing and trace a bit, split/merge a tree, delete a tree, whatever.
  - Then open the version restore view by clicking `Restore older version` in the dropdown next to the save button. The tracing should be saved and in read only mode now.
  - Preview any version by selecting it. You should not be able to modify the tracing.
  - Now there are two options:
    - Either close the version restore view. The newest version should be loaded again and you should be able to modify it and save.
    - Or select `Restore this version` for an older version. The old version should be loaded and you should be able to modify it and save. Opening the version restore view again should show that you reverted to a previous version.
- Create a volume tracing - the option to open the version restore view should not be visible.

### Issues:
- fixes #51 :open_mouth: 

------
